File: apps/mobile/android/app/src/main/kotlin/com/ilkersevim/blocflutter/NativeShowcaseTelemetryStreamHandler.kt
package com.ilkersevim.blocflutter

import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import io.flutter.plugin.common.EventChannel

class NativeShowcaseTelemetryStreamHandler : EventChannel.StreamHandler {
  private var workerThread: HandlerThread? = null
  private var workerHandler: Handler? = null
  private var mainHandler: Handler? = null
  private var eventSink: EventChannel.EventSink? = null
  private var sampleRunnable: Runnable? = null
  private var emitRunnable: Runnable? = null
  private var sessionGeneration = 0L
  private var accumulator: NativeShowcaseTelemetryAccumulator? = null

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    stopSession()

    val parsed = NativeShowcaseTelemetryAccumulator.parseConfig(arguments)
    if (parsed == null) {
      mainHandler = Handler(Looper.getMainLooper())
      mainHandler?.post {
        events?.error(
          "invalid_config",
          "Telemetry stream config must be schemaVersion=1 render mode with valid fields.",
          null,
        )
      }
      return
    }

    sessionGeneration += 1
    val generation = sessionGeneration
    val localAccumulator =
      NativeShowcaseTelemetryAccumulator(
        sessionId = parsed.sessionId,
        aggregation = parsed.aggregation,
        deliveredRateHz = parsed.deliveredRateHz,
      )
    accumulator = localAccumulator

    // Sink handoff must stay on the platform main thread even when onListen
    // runs on a background task queue.
    mainHandler = Handler(Looper.getMainLooper())
    mainHandler?.post {
      if (generation == sessionGeneration) {
        eventSink = events
      }
    }

    workerThread = HandlerThread("NativeShowcaseTelemetry").also { it.start() }
    workerHandler = Handler(workerThread!!.looper)

    val sampleIntervalMs =
      (1000.0 / NativeShowcaseTelemetryAccumulator.SOURCE_RATE_HZ)
        .toLong()
        .coerceAtLeast(1L)
    val emitIntervalMs = parsed.deliveryWindowMs

    sampleRunnable =
      object : Runnable {
        override fun run() {
          if (generation != sessionGeneration) {
            return
          }
          localAccumulator.ingest(localAccumulator.demoSampleValue())
          workerHandler?.postDelayed(this, sampleIntervalMs)
        }
      }

    emitRunnable =
      object : Runnable {
        override fun run() {
          if (generation != sessionGeneration) {
            return
          }
          val payload = localAccumulator.emitIfReady()
          if (payload != null) {
            mainHandler?.post {
              if (generation == sessionGeneration) {
                eventSink?.success(payload)
              }
            }
          }
          workerHandler?.postDelayed(this, emitIntervalMs)
        }
      }

    workerHandler?.post(sampleRunnable!!)
    workerHandler?.postDelayed(emitRunnable!!, emitIntervalMs)
  }

  override fun onCancel(arguments: Any?) {
    stopSession()
  }

  private fun stopSession() {
    sessionGeneration += 1
    sampleRunnable?.let { workerHandler?.removeCallbacks(it) }
    emitRunnable?.let { workerHandler?.removeCallbacks(it) }
    sampleRunnable = null
    emitRunnable = null
    workerHandler = null
    workerThread?.quitSafely()
    workerThread = null
    mainHandler?.post {
      eventSink = null
    }
    mainHandler = null
    accumulator?.resetSession()
    accumulator = null
  }
}

File: apps/mobile/ios/Runner/NativeShowcaseTelemetryStreamHandler.swift
import Flutter
import Foundation

final class NativeShowcaseTelemetryStreamHandler: NSObject, FlutterStreamHandler {
  private let workerQueue = DispatchQueue(label: "com.example.flutter_bloc_app.telemetry.worker")
  private var emitTimer: DispatchSourceTimer?
  private var sampleTimer: DispatchSourceTimer?
  private var eventSink: FlutterEventSink?
  private var sessionGeneration: UInt64 = 0
  private var accumulator: NativeShowcaseTelemetryAccumulator?

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    guard let parsed = NativeShowcaseTelemetryAccumulator.parseConfig(arguments) else {
      return FlutterError(
        code: "invalid_config",
        message: "Telemetry stream config must be schemaVersion=1 render mode with valid fields.",
        details: nil
      )
    }

    workerQueue.async { [weak self] in
      guard let self else { return }
      self.tearDownSessionOnWorkerQueue()
      self.sessionGeneration &+= 1
      let generation = self.sessionGeneration
      let localAccumulator = NativeShowcaseTelemetryAccumulator(
        sessionId: parsed.sessionId,
        aggregation: parsed.aggregation,
        deliveredRateHz: parsed.deliveredRateHz
      )
      localAccumulator.resetSession()
      self.accumulator = localAccumulator
      DispatchQueue.main.async { [weak self] in
        guard let self, generation == self.sessionGeneration else { return }
        self.eventSink = events
      }
      self.startWorkers(generation: generation, config: parsed, accumulator: localAccumulator)
    }
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    workerQueue.async { [weak self] in
      guard let self else { return }
      self.sessionGeneration &+= 1
      self.tearDownSessionOnWorkerQueue()
      self.accumulator?.resetSession()
      self.accumulator = nil
      DispatchQueue.main.async { [weak self] in
        self?.eventSink = nil
      }
    }
    return nil
  }

  private func startWorkers(
    generation: UInt64,
    config: NativeShowcaseTelemetryAccumulator.ParsedConfig,
    accumulator: NativeShowcaseTelemetryAccumulator
  ) {
    let sampleInterval = DispatchTimeInterval.nanoseconds(
      Int(1_000_000_000 / NativeShowcaseTelemetryAccumulator.sourceRateHz)
    )
    let emitInterval = DispatchTimeInterval.milliseconds(config.deliveryWindowMs)

    let sampleTimer = DispatchSource.makeTimerSource(queue: workerQueue)
    sampleTimer.schedule(deadline: .now(), repeating: sampleInterval)
    sampleTimer.setEventHandler { [weak self] in
      guard let self, generation == self.sessionGeneration else { return }
      accumulator.ingest(sampleValue: accumulator.demoSampleValue())
    }
    sampleTimer.resume()
    self.sampleTimer = sampleTimer

    let emitTimer = DispatchSource.makeTimerSource(queue: workerQueue)
    emitTimer.schedule(deadline: .now() + emitInterval, repeating: emitInterval)
    emitTimer.setEventHandler { [weak self] in
      guard let self, generation == self.sessionGeneration else { return }
      guard let payload = accumulator.emitIfReady() else { return }
      DispatchQueue.main.async { [weak self] in
        guard let self, generation == self.sessionGeneration else { return }
        self.eventSink?(payload)
      }
    }
    emitTimer.resume()
    self.emitTimer = emitTimer
  }

  private func tearDownSessionOnWorkerQueue() {
    sampleTimer?.cancel()
    sampleTimer = nil
    emitTimer?.cancel()
    emitTimer = nil
  }
}
